### PR TITLE
[move-compiler v2] Comparison testing for file-format generator

### LINKED ISSUE #9334: [Bug][compiler v2] Type inference for unbound lets
In

```
script {
fun main() {
    let x;
    if (true)
        x = 3
    else
        x = 5;
    x;
}
}

```

The compiler infers incorrectly the unit type, and then runs in a follow-up bug later in the pipeline:
```
 error: cannot assign tuple type `()` to single variable (use `(a, b, ..) = ..` instead)
  ┌─ /var/folders/3c/0ypx5xvd7v1078zhfjw16q8r0000gn/T/.tmpaQ34By:4:9
  │
4 │     let x;
  │         ^

bug: inconsistent numeric constant: Tuple([])
  ┌─ /var/folders/3c/0ypx5xvd7v1078zhfjw16q8r0000gn/T/.tmpaQ34By:6:13
  │
6 │         x = 3
  │             ^

bug: inconsistent numeric constant: Tuple([])
  ┌─ /var/folders/3c/0ypx5xvd7v1078zhfjw16q8r0000gn/T/.tmpaQ34By:8:13
  │
8 │         x = 5;
  │             ^

bug: inconsistent expression target arity
  ┌─ /var/folders/3c/0ypx5xvd7v1078zhfjw16q8r0000gn/T/.tmpaQ34By:9:5
  │
9 │     x;
  │     ^

```

### PR DESCRIPTION
**Pull request by wrwg**
This extends the transactional-test-runner to support compiler V2 _and_ to do comparison testing between V1 and V2.

It ports over most of the transactional tests of the V1 compiler into the V2 tree and runs comparison testing on them, fixing bugs throughout the compilation chain as needed.

Closes #9334

NOTE: this is diffbased on #9258; only look at the latest commit.

### CODE REVIEW

**Review by rahxephon89 — APPROVED** [review]

File: third_party/move/move-compiler-v2/transactional-tests/tests/evaluation_order/struct_arguments.exp [hunk 1]
```diff
@@ -0,0 +1,69 @@
+comparison between v1 and v2 failed:
```
Comment (vgao1996) [inline third_party/move/move-compiler-v2/transactional-tests/tests/evaluation_order/struct_arguments.exp:1]: I assume you'll fix these failed tests at a later stage, right?

**Review by vgao1996 — APPROVED** [review]

File: third_party/move/move-compiler-v2/transactional-tests/tests/evaluation_order/struct_arguments.exp (hunk 1 above)
Comment (wrwg) [reply third_party/move/move-compiler-v2/transactional-tests/tests/evaluation_order/struct_arguments.exp:1]: Yes! https://github.com/aptos-labs/aptos-core/issues/9343

File: third_party/move/move-command-line-common/src/testing.rs [hunk 2]
```diff
@@ -63,3 +63,29 @@ pub fn format_diff(expected: impl AsRef<str>, actual: impl AsRef<str>) -> String
     }
     ret
 }
+
+pub fn format_diff_no_color(expected: impl AsRef<str>, actual: impl AsRef<str>) -> String {
```
Comment (brmataptos) [inline third_party/move/move-command-line-common/src/testing.rs:67]: This is kind of redundant with the code above, but I can't simplify it much in Rust, so it's fine as is.  Higher-order functions in Rust are a bit verbose.

Here's a try:
```
pub fn format_diff_no_color(expected: impl AsRef<str>, actual: impl AsRef<str>) -> String {
    format_diff_shared(
        expected,
        actual,
        |x: &String, ret: &mut String| {
            ret.push_str(&format!("= {}", x.replace('\n', "\n= ")));
            ret.push('\n');
        },
        |x: &String, ret: &mut String| {
            ret.push_str(&format!("+ {}", x.replace('\n', "\n+ ")));
            ret.push('\n');
        },
        |x: &String, ret: &mut String| {
            ret.push_str(&format!("- {}", x.replace('\n', "\n- ")));
            ret.push('\n');
        },
    )
}

```

Where the code above is changed to:
```

fn format_diff_shared(
    expected: impl AsRef<str>,
    actual: impl AsRef<str>,
    same_proc: impl Fn(&String, &mut String),
    add_proc: impl Fn(&String, &mut String),
    rem_proc: impl Fn(&String, &mut String),
) -> String {
    use difference::{Changeset, Difference};

    let changeset = Changeset::new(expected.as_ref(), actual.as_ref(), "\n");

    let mut ret = String::new();

    for seq in changeset.diffs {
        match &seq {
            Difference::Same(x) => {
                same_proc(x, &mut ret);
            },
            Difference::Add(x) => {
                add_proc(x, &mut ret);
            },
            Difference::Rem(x) => {
                rem_proc(x, &mut ret);
            },
        }
    }
    ret
}

pub fn format_diff(expected: impl AsRef<str>, actual: impl AsRef<str>) -> String {
    format_diff_shared(
        expected,
        actual,
        |x: &String, ret: &mut String| {
            ret.push_str(x);
            ret.push('\n');
        },
        |x: &String, ret: &mut String| {
            ret.push_str("\x1B[92m");
            ret.push_str(x);
            ret.push_str("\x1B[0m");
            ret.push('\n');
        },
        |x: &String, ret: &mut String| {
            ret.push_str("\x1B[91m");
            ret.push_str(x);
            ret.push_str("\x1B[0m");
            ret.push('\n');
        },
    )
}
```

**Review by brmataptos — COMMENTED** [review]
Hopefully someone looked at the tests better than I.

**Pull request merged**